x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,275 @@
+
+	packet, err := ibctesting.ParsePacketFromEvents(res.GetEvents())
+	suite.Require().NoError(err)
+
+	err = suite.path.EndpointA.UpdateClient()
+	suite.Require().NoError(err)
+
+	res, err = suite.path.EndpointA.RecvPacketWithResult(packet)
+	suite.Require().NoError(err)
+
+	ack, err := ibctesting.ParseAckFromEvents(res.GetEvents())
+	return string(ack), err
+}
+
+func (suite *MiddlewareTestSuite) AssertReceiveSuccess(success bool, msg sdk.Msg) (string, error) {
+	ack, err := suite.ExecuteReceive(msg)
+	if success {
+		suite.Require().NoError(err)
+		suite.Require().NotContains(string(ack), "error",
+			"acknoledgment is an error")
+	} else {
+		suite.Require().Contains(string(ack), "error",
+			"acknoledgment is not an error")
+		suite.Require().Contains(string(ack), types.RateLimitExceededMsg,
+			"acknoledgment error is not of the right type")
+	}
+	return ack, err
+}
+
+func (suite *MiddlewareTestSuite) AssertSendSuccess(success bool, msg sdk.Msg) (*sdk.Result, error) {
+	r, err := suite.chainA.SendMsgsNoCheck(msg)
+	if success {
+		suite.Require().NoError(err, "IBC send failed. Expected success. %s", err)
+	} else {
+		suite.Require().Error(err, "IBC send succeeded. Expected failure")
+		suite.ErrorContains(err, types.RateLimitExceededMsg, "Bad error type")
+	}
+	return r, err
+}
+
+func (suite *MiddlewareTestSuite) BuildChannelQuota(name string, duration, send_precentage, recv_percentage uint32) string {
+	return fmt.Sprintf(`
+          {"channel_id": "channel-0", "denom": "%s", "quotas": [{"name":"%s", "duration": %d, "send_recv":[%d, %d]}] }
+    `, sdk.DefaultBondDenom, name, duration, send_precentage, recv_percentage)
+}
+
+// Tests
+
+// Test that Sending IBC messages works when the middleware isn't configured
+func (suite *MiddlewareTestSuite) TestSendTransferNoContract() {
+	one := sdk.NewInt(1)
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, one))
+}
+
+// Test that Receiving IBC messages works when the middleware isn't configured
+func (suite *MiddlewareTestSuite) TestReceiveTransferNoContract() {
+	one := sdk.NewInt(1)
+	suite.AssertReceiveSuccess(true, suite.NewValidMessage(false, one))
+}
+
+// Test rate limiting on sends
+func (suite *MiddlewareTestSuite) TestSendTransferWithRateLimiting() map[string]string {
+	// Setup contract
+	suite.chainA.StoreContractCode(&suite.Suite)
+	quotas := suite.BuildChannelQuota("weekly", 604800, 5, 5)
+	addr := suite.chainA.InstantiateContract(&suite.Suite, quotas)
+	suite.chainA.RegisterRateLimitingContract(addr)
+
+	// Setup sender chain's quota
+	osmosisApp := suite.chainA.GetOsmosisApp()
+
+	// Each user has 10% of the supply
+	supply := osmosisApp.BankKeeper.GetSupply(suite.chainA.GetContext(), sdk.DefaultBondDenom)
+	quota := supply.Amount.QuoRaw(20)
+	half := quota.QuoRaw(2)
+
+	// send 2.5% (quota is 5%)
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, half))
+
+	// send 2.5% (quota is 5%)
+	r, _ := suite.AssertSendSuccess(true, suite.NewValidMessage(true, half))
```

Comment:
err check

Reply:
This is already checked inside the call so it matches the first param. I can check it again, but we'd be duplicating logic. 

The reason for returning it is that in some tests we may want to do more checks (the error type, for example). Here it's useful to get the response so we can check the attrs, but in most cases we can just call it and ignore the results.

Reply:
I see, makes sense. I still think handling errors any time they are returned is better in case the callee logic changes. Not feeling strongly about this though so feel free to resolve if you disagree

Reply:
I generally would agree, but since we want to check different type of errors it was easier to encapsulate it.

To be fair, I've had to override some of the ibctesting functions for this exact reason: they assert NoError inside the function and I want to ignore failures (cause failures is what I'm testing). 

I'll see if there's a clean way to handle these outside and otherwise resolve this